### PR TITLE
awsbase: Add unit testing for session code base

### PR DESCRIPTION
**Pull request by nywilken**
The changes within this branch are a fast past at extending the testing with the awsbase pkg. In its present state it adds some basic checks for the GetSession* logic using some of the existing AWS mocks. During this exercise I've started work in looking through the tests and code structure to determine how authentication is flowing. While looking for any special edge cases or chaining that is occurring to better understand the coverage of the authentication testing already in place. This PR is merely a starting step, and anyone reviewing this code is welcome to provide feedback, improvement suggestions.

Relates #4 
Relates #5
Relates #7 

Goals:
- [ ] Test all AWS supported GetSession use cases
- [ ] Test AWS authentication precedence as documented for the AWS Terraform Provider
- [x] Enable Travis CI on master and new pull-requests (`make lint && make test`)

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -33,40 +33,7 @@ func GetSessionOptions(c *Config) (*session.Options, error) {
 		return nil, err
 	}
 
-	// Call Get to check for credential provider. If nothing found, we'll get an
```
Comment (nywilken) [inline session.go]: Looking through the chain of events there appears to be an error check for the "NoCredentialProviders" error within the call to GetCredentials which then gets bubbled up here. So I removed this call for now and validated that with the current testing in place nothing breaks. This still needs some additional validation :man_shrugging:

File: session.go [hunk 2]
```diff
@@ -180,15 +155,14 @@ func GetSessionWithAccountIDAndPartition(c *Config) (*session.Session, string, s
 		}
 
 		accountID, partition, err := GetAccountIDAndPartition(iamClient, stsClient, credentialsProviderName)
-
-		if err == nil {
-			return sess, accountID, partition, nil
+		if err != nil {
```
Comment (nywilken) [inline session.go]: Happy path - return if there is an error. Otherwise continue happily down the path of normal execution.

File: session_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.desc, func(t *testing.T) {
+			opts, err := GetSessionOptions(tc.config)
+			if err != nil && tc.hasError == false {
+				t.Fatalf("GetSessionOptions(c) resulted in an error %s", err)
+			}
+
+			if opts == nil && tc.hasError == false {
```
Comment (nywilken) [inline session_test.go]: Open question: What is the best way to validate the returned session?

File: awsauth_test.go [hunk 4]
```diff
@@ -420,25 +418,27 @@ func TestAWSParseAccountIDAndPartitionFromARN(t *testing.T) {
...
 	defer resetEnv()
-	cfg := Config{}
 
+	cfg := Config{}
 	c, err := GetCredentials(&cfg)
-	if awsErr, ok := err.(awserr.Error); ok {
-		if awsErr.Code() != "NoCredentialProviders" {
-			t.Fatal("Expected NoCredentialProviders error")
+
+	if err == nil {
```
Comment (nywilken) [inline awsauth_test.go]: The logic within this branch seems like the normal flow of execution so I would be inclined to breaking it out of the if/else condition and instead branch if `err != nil`. In Go, you will find that else statements are not used that often because the community tends to lean more towards only using the if statement as guard clauses for conditions that would break the normal flow of execution. For example, if `err != nil` then exit the test. 

Here's a blog post that may helpful better explain this comment https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88
